codelabs/codelab-02/codelab2README.md

```diff
@@ -0,0 +1,501 @@
+
+Great, you should now be familiar with the basic operations on S3. You can programmatically create buckets, upload objects, read back data from objects, list keys in a bucket, and delete these objects by key.
+
+#### Storage Types
+
+In class, we talked about four storage classes: Default, Reduced Redundancy Storage (RRS), Infrequently Accessed (IA), and Glacier. We're not going to cover Glacier here, since it's API is fairly complicated. However, let's experiment with moving objects between the other three storage types.
+
+We can toggle between the first three storage classes of an S3 object using the `—storage-class` flag for the `s3api put-object` command. Note that moving to Glacier is a separate process. Why? Objects in the first three classes are all accessible in real-time, while objects in Glacier are not.
+
+For this process, let's re-upload a file into our bucket:
+
+```
+$ aws s3api put-object —bucket cmsc389l-colink —key hello_world.txt —body hello_world.txt
+```
+
+##### Configuring RRS or IA
+
+Before we change the storage class, we can inspect the contents of the file we just uploaded to determine what storage class it is held in.
+
+```
+$ aws s3api list-objects —bucket cmsc389l-colink —prefix hello_world.txt
+```
+
+There are two ways to modify the storage class. The first involves overwriting the object already in S3 with a new object which has a different storage class set:
+
+```
+$ aws s3api put-object —bucket cmsc389l-colink —key hello_world.txt —storage-class REDUCED_REDUNDANCY
+```
+
+Perform the `s3api list-objects` command again to double-check that it modified the storage class.
+
+The second is that we can use the `s3api copy-object` command to avoid the extra bandwidth charges that arise from uploading the object all over again. In this case, it will only change the storage class on the object.
+
+```
+$ aws s3api copy-object —bucket cmsc389l-colink —copy-source cmsc389l-colink/hello_world.txt —key hello_world.txt —storage-class STANDARD_IA
+```
+
+If you perform one last `s3api list-objects` command, you'll see that it changed the storage class again.
+
+Remember that each of these modes still provide real-time access to these objects, so we can read them and get them back quickly:
+
+```
+$ time aws s3api get-object —bucket cmsc389l-colink —key hello_world.txt out.txt
+...
+aws s3api get-object —bucket cmsc389l-colink —key hello_world.txt out.txt 0.29s user 0.05s system 41% cpu 0.817 total
+```
+
+Instead, if you were to move these files into Glacier, then you would need to trigger a transfer job which would move files back into an S3 bucket. This process will usually take hours.
+#### S3 CLI
+
+The tool that you've used so far, `aws s3api`, is one of two AWS CLI tools for interacting with S3. The other, `aws s3`, is a higher-level wrapper on top of `s3api` that provides typical directory traversal operations.
+
+For example, you can `ls` a directory on S3:
+
+```
+$ aws s3 ls s3://cmsc389l-colink
```

Comment:
Should be indented

https://umd-cs-stics.gitbooks.io/cmsc389l-fall2017/content/lectures/lecture-02/codelab.html#s3-cli